File: Solutions/111-Minimum-Depth-of-Binary-Tree/Solution2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int minDepth(TreeNode* root) {
        // BFS
        int depth = 0;
        queue<TreeNode*> levelNode;
        if(root) levelNode.push(root);
        while(!levelNode.empty()){
            depth++;
            int n_nodes = levelNode.size();
            for(int i = 0; i < n_nodes; i++){
                TreeNode* tmp = levelNode.front();
                if(!tmp->left && !tmp->right)
                    return depth;
                if(tmp->left)
                    levelNode.push(tmp->left);
                if(tmp->right)
                    levelNode.push(tmp->right);
                levelNode.pop();
            }
        }
        return depth;
    }
};

File: Solutions/103-Binary-Tree-Zigzag-Level-Order-Traversal/Solution2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
from collections import deque
class Solution:
    def zigzagLevelOrder(self, root: TreeNode) -> List[List[int]]:
        vals = []
        if not root:
            return vals
        lvl_vals, lvl_nodes = [], deque()
        lvl_nodes.append(root)
        depth = 0
        while lvl_nodes:
            n_nodes = len(lvl_nodes)
            for _ in range(n_nodes):
                node = lvl_nodes.popleft()
                lvl_vals.append(node.val)
                if node.left:
                    lvl_nodes.append(node.left)
                if node.right:
                    lvl_nodes.append(node.right)
            if lvl_vals:
                if depth % 2 != 0:
                    lvl_vals.reverse() 
                vals.append(lvl_vals)
                lvl_vals= []
                depth += 1
                
            
        return vals
        
        
        
        
